File: Java/PuntoVentaDulces/src/Clases/GestorInventario.java
package Clases;

import java.io.*;
import java.util.ArrayList;
import java.util.function.Function;

public class GestorInventario<T extends Producto> {
    private ArrayList<T> lista;
    private final String rutaArchivo;
    
    private final Function<String, T> conversorDesdeLinea;
    private final Function<T, String> conversorALinea;

    public GestorInventario(String rutaArchivo, Function<String, T> desdeLinea, Function<T, String> aLinea) {
        this.lista = new ArrayList<>();
        this.rutaArchivo = rutaArchivo;
        this.conversorDesdeLinea = desdeLinea;
        this.conversorALinea = aLinea;
        cargarDatos();
    }

    public void agregar(T elemento) {
        lista.add(elemento);
        guardarDatos();
    }

    public ArrayList<T> obtenerTodos() {
        return lista;
    }

    public T buscarPorCodigo(String codigo) throws InventarioException {
        for (T elemento : lista) {
            if (elemento.getCodigo().equals(codigo)) {
                return elemento;
            }
        }
        throw new InventarioException("El código '" + codigo + "' no existe en la base de datos.");
    }

    public void sincronizarCambios() {
        guardarDatos();
    }

    public void eliminar(T elemento) {
        lista.remove(elemento);
        guardarDatos();
    }

    private void cargarDatos() {
        File archivo = new File(rutaArchivo);
        if (!archivo.exists()) return;
        try (BufferedReader br = new BufferedReader(new FileReader(archivo))) {
            String linea;
            while ((linea = br.readLine()) != null) {
                T entidad = conversorDesdeLinea.apply(linea);
                if (entidad != null) lista.add(entidad);
            }
        } catch (Exception e) {
            System.err.println("Error al cargar los datos.");
        }
    }

    private void guardarDatos() {
        try (PrintWriter pw = new PrintWriter(new FileWriter(rutaArchivo))) {
            for (T e : lista) {
                pw.println(conversorALinea.apply(e));
            }
        } catch (Exception e) {
            System.err.println("Error al guardar.");
        }
    }
}

File: PuntoVentaDulces/PuntoVentaDulces/GestorInventario.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace Clases
{
    public class GestorInventario<T> where T : Producto
    {
        private List<T> lista;
        private readonly string rutaArchivo;

        private readonly Func<string, T> conversorDesdeLinea;
        private readonly Func<T, string> conversorALinea;

        public GestorInventario(string rutaArchivo, Func<string, T> desdeLinea, Func<T, string> aLinea)
        {
            this.lista = new List<T>();
            this.rutaArchivo = rutaArchivo;
            this.conversorDesdeLinea = desdeLinea;
            this.conversorALinea = aLinea;
            CargarDatos();
        }

        public void Agregar(T elemento)
        {
            lista.Add(elemento);
            GuardarDatos();
        }

        public List<T> ObtenerTodos()
        {
            return lista;
        }

        public T BuscarPorCodigo(string codigo)
        {
            foreach (T elemento in lista)
            {
                if (elemento.GetCodigo() == codigo)
                {
                    return elemento;
                }
            }
            throw new InventarioException("El codigo '" + codigo + "' no existe en la base de datos.");
        }

        public void SincronizarCambios()
        {
            GuardarDatos();
        }

        public void Eliminar(T elemento)
        {
            lista.Remove(elemento);
            GuardarDatos();
        }

        private void CargarDatos()
        {
            if (!File.Exists(rutaArchivo)) return;

            try
            {
                using (StreamReader br = new StreamReader(rutaArchivo))
                {
                    string linea;
                    while ((linea = br.ReadLine()) != null)
                    {
         
                        T entidad = conversorDesdeLinea(linea);
                        if (entidad != null) lista.Add(entidad);
                    }
                }
            }
            catch (Exception)
            {
                Console.Error.WriteLine("Error al cargar los datos.");
            }
        }

        private void GuardarDatos()
        {
            try
            {
                using (StreamWriter pw = new StreamWriter(rutaArchivo))
                {
                    foreach (T e in lista)
                    {
                        pw.WriteLine(conversorALinea(e));
                    }
                }
            }
            catch (Exception)
            {
                Console.Error.WriteLine("Error al guardar.");
            }
        }
    }
}